provider/mesh/state.go

```diff
@@ -70,6 +52,47 @@ func (st *notificationState) gc(retention time.Duration) {
 	}
 }
 
+func (st *notificationState) snapshot(w io.Writer) error {
```

Comment:
See above, no relation between gossip state and snapshot encoding.

I'm not 100% sure how gob decodes a map internally, but I imagine it is not streaming as we are doing here. That'll at least make recovering partially corrupted snapshots hard at some point.
Separation of key/value is only because that's how the current types are.

Material for later PRs to not blow up this one.